### 1)Загружаем необходимые библиотеки

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats

###2) Подготовка данных:

* Загрузка данных из моего репозитория GitHub.
* Переименовываем столбец 'Unnamed: 0' в 'date'
* Приводем данные столбца 'date' к типу дата и делаем его индексом

In [ ]:
data = pd.read_csv('https://raw.githubusercontent.com/gainutdinov-rafael/architectAI/main/certification1/dataset_tk.csv')
data = data.rename(columns={'Unnamed: 0':'date'})
data['date'] = pd.to_datetime(data['date'], format='%d/%m/%Y %H:%M:%S')
data.set_index(['date'],inplace=True)
data.head(1000)

Punjab  Haryana  Rajasthan  Delhi     UP  Uttarakhand    HP   J&K  \
date                                                                            
2019-01-02   119.9    130.3      234.1   85.8  313.9         40.7  30.0  52.5   
2019-01-03   121.9    133.5      240.2   85.5  311.8         39.3  30.1  54.1   
2019-01-04   118.8    128.2      239.8   83.5  320.7         38.1  30.1  53.2   
2019-01-05   121.0    127.5      239.1   79.2  299.0         39.2  30.2  51.5   
2019-01-06   121.4    132.6      240.4   76.6  286.8         39.2  31.0  53.2   
...            ...      ...        ...    ...    ...          ...   ...   ...   
2020-12-01   130.8    126.0      220.4   77.3  322.8         36.5  25.4  45.4   
2020-12-02   129.4    127.3      218.4   81.3  331.4         37.0  25.8  45.2   
2020-12-03   132.1    129.7      205.6   85.6  336.7         37.8  26.2  45.2   
2020-12-04   132.1    128.4      207.0   83.9  334.6         38.2  27.0  43.0   
2020-12-05   120.4    111.2      201.4   83.6  287.3         35.2  24.3  42.6   

            Chandigarh  Chhattisgarh  ...  Odisha  West Bengal  Sikkim  \
date                                  ...                                
2019-01-02         5.0          78.7  ...    70.2        108.2     2.0   
2019-01-03         4.9          78.8  ...    67.9        110.2     1.9   
2019-01-04         4.8          74.8  ...    66.3        106.8     1.7   
2019-01-05         4.3          69.0  ...    65.8        107.0     2.0   
2019-01-06         4.3          68.1  ...    62.9        106.4     2.0   
...                ...           ...  ...     ...          ...     ...   
2020-12-01         3.6         100.4  ...   105.7        172.6     1.4   
2020-12-02         3.7         101.8  ...   112.3        174.4     1.5   
2020-12-03         4.0         103.0  ...   113.0        169.8     1.4   
2020-12-04         4.0         101.8  ...   111.6        145.5     1.2   
2020-12-05         3.5         102.4  ...    95.1        110.4     1.2   

            Arunachal Pradesh  Assam  Manipur  Meghalaya  Mizoram  Nagaland  \
date                                                                          
2019-01-02                2.1   21.7      2.7        6.1      1.9       2.2   
2019-01-03                2.2   23.4      2.4        6.5      1.8       2.2   
2019-01-04                2.2   21.7      2.4        6.3      1.7       2.2   
2019-01-05                2.2   22.5      2.7        5.7      1.8       2.3   
2019-01-06                2.2   21.7      2.7        6.2      1.9       2.3   
...                       ...    ...      ...        ...      ...       ...   
2020-12-01                2.1   20.4      2.1        5.4      1.8       2.0   
2020-12-02                2.1   23.3      2.4        5.4      1.6       2.1   
2020-12-03                2.1   21.7      2.3        5.4      1.5       2.0   
2020-12-04                2.0   24.2      2.5        5.7      1.6       2.0   
2020-12-05                2.1   20.3      2.5        5.8      1.6       2.1   

            Tripura  
date                 
2019-01-02      3.4  
2019-01-03      3.6  
2019-01-04      3.5  
2019-01-05      3.5  
2019-01-06      3.3  
...             ...  
2020-12-01      2.4  
2020-12-02      3.3  
2020-12-03      3.6  
2020-12-04      3.3  
2020-12-05      3.3  

[503 rows x 33 columns]

###3) Дополнение таблица необходимыми данными

* добавим и расчитаем столбец общего потребления электроэнергии по всей стране
* добавим и расчитаем среднее потребление электтроэнергии по всей стране

In [ ]:
data = data.assign(sum=0)
data = data.assign(mean=0)
for col in data.columns:
  data['sum']=data['sum']+data[col]
data['mean'] = data['sum']/len(data.columns)

###4) Ищем выбросы

В статистике выбросы — это значения, резко отличающиеся от других значений в собранном наборе данных. Выброс может указывать на аномалии в распределении данных или на ошибки при измерениях, поэтому зачастую выбросы исключаются из набора данных.

###Графики для поиска выбросов по группам

Группы по потреблению: первый график - 0-10 гВт, второй - 10-50 гВт, третий - 50-500 гВт

In [ ]:
sns.boxplot(data[['Punjab','Haryana','Rajasthan','Delhi','UP','Gujarat','MP','Maharashtra','Andhra Pradesh','Telangana','Karnataka','Tamil Nadu','Bihar','Odisha','West Bengal']],width=0.8,orient='h')

In [ ]:
sns.boxplot(data[['Uttarakhand','HP','J&K','Goa','DNH','Jharkhand','Assam']],width=0.8,orient='h')

In [ ]:
sns.boxplot(data[['Chandigarh','Pondy','Sikkim','Arunachal Pradesh','Manipur','Meghalaya','Mizoram','Nagaland','Tripura']],width=0.8,orient='h')

###5) Из графиков видно большое колиество выбросов в некоторых штатах. Для их отчистки применим 2 способа:
* Очистка методом: z-оценка
* Очистка методом межквартильного диапазона

In [ ]:
z = np.abs(stats.zscore(data))
data_clean1 = data[(z<3).all(axis=1)]
data_clean1.shape

In [ ]:
Q1 = data.quantile(q=.25)
Q3 = data.quantile(q=.75)
IQR = data.apply(stats.iqr)
data_clean2 = data[~((data < (Q1-1.5*IQR)) | (data > (Q3+1.5*IQR))).any(axis=1)]
data_clean2.shape

В первоначальном наборе даннных было 503 записи. При очистке по z-оценке осталось 440 записей, а при очистке методом межквартильного диапазона осталось 249 записей.

Ни один из способов не является 100% верным, т.к. нельзя однозначно сказать является то или иное значение ошибочным, но исключая выбросы можно получить новую информацию.

###График суммарного потребления электроэнергии штатами в Индии

график суммарного потребления нужен чтобы видеть "картину экономики в целом".

In [ ]:
fig,axes = plt.subplots(nrows=1)
fig.suptitle('График суммарного потребления энергии',fontsize=22)
data['sum'].plot(x=data.index,y=data['sum'],figsize=(30,10),label='До очистки',ax=axes,lw=5,linestyle='-',marker='o',markerfacecolor='black')
data_clean1['sum'].plot(x=data_clean1.index,y=data_clean1['sum'],figsize=(30,10),label='Очистка z-оценкой',ax=axes,lw=5,linestyle=':',marker='v',markerfacecolor='red')
data_clean2['sum'].plot(x=data_clean2.index,y=data_clean2['sum'],figsize=(30,10),label='Очистка методом межквартильного диапазона',ax=axes,lw=5,linestyle=':',marker='o',markerfacecolor='blue')

axes.legend(fontsize=12)
axes.grid(linestyle='--', alpha=0.8,axis='y')

#data['sum'].plot(x=data.index,figsize=(30,10))

Глядя на график можно сделать вывод, что что то произошло в июне 2020 года, после чего график стал "ступенчатым". Если вспомнить события 2020 года, то можно предположить, что причиной такого изменения стала пандемия Ковид2019.

На следующем графике сравним потребление двумя случайными штатами и сравним его со средними знаениями.

In [ ]:
fig,axes = plt.subplots(nrows=1)
data['UP'].plot(x=data.index,figsize=(30,10),label='UP')
data['Maharashtra'].plot(x=data.index,figsize=(30,10),label='Maharashtra')
data['mean'].plot(x=data.index,figsize=(30,10),label='Среднее значение')
plt.legend(fontsize=12)
plt.show()

По графику можно пронаблюдать периодическое расхождение в "разные стороны", т.е. когда в одном штате потребление растет, в другом потребление электроэнергии падает. Махараштра — лидирующий промышленный штат страны, Уттар-Прадеш — один из наиболее отсталых в экономическом отношении штатов страны. В то же время, Уттар Прадеш — вторая в стране экономика (после Махараштры).

В период пандемии Ковид19 в сентябре-октябре 2020 года. Махараштра резко увеличила потребление электоэнергии, что может говорить о запуске промышленности, а Уттар-Прадеш наоборот сильно упало потребление, что является признаком падения производства как основного потребителя электроэнергии. Т.е. Махараштра является более экономически устойчивым штатом чем Уттар-Прадеш.

Добавим график зависомости потребления электроэнергии штта от среднего значения потребления электроэнергии по всем штатам, для того чтобы посмотреть на аномалии.

In [ ]:
fig,axes = plt.subplots()
#axes.scatter(data['mean'], data['Pondy'],label='Pondy')
axes.scatter(data['mean'], data['Sikkim'],label='Sikkim')
axes.scatter(data['mean'], data['Tripura'],label='Tripura')
axes.scatter(data['mean'], data['Chandigarh'],label='Chandigarh')
plt.legend(fontsize=12)
plt.show()

по графику видно что иногда встречаются аномалии, но в целом разброс не значительный.

Попробуем посмотреть на какие периоды выпадало наибольшее потребление электроэнергии. Оценку будем проводить по среднему значению.

In [ ]:
data.where(data['mean'] > 230).dropna().head()

По полученным данным видно, что наибольшее потребление приходилось на февраль 2019 года. Т.е. после начала пандемии произошло общее падение производства.

Графики потребления по всем штатам. 35 графиков в одном размещать не удобно, поэтому сделал разделение на несколько графиков. Через переменную **ax** можно установить необходимое количество графиков для разбивки.

In [ ]:
ax = 3 #Количество графиков
fig,axes = plt.subplots(nrows=ax)
fig.suptitle('График потребления энергии по штатам',fontsize=22)
fig.set_dpi(1000)
i=0
for col in data:
  if col != 'sum':
    data[col].plot(x=data.index,y=data[col],figsize=(30,30),label=col,ax=axes[i%ax])
    i += 1

for i in range(0,ax):
  axes[i].legend(fontsize=12)
  axes[i].grid(linestyle='--', alpha=0.8,axis='y')


#Выводы

В результате проделанной работы были загружены с GitHub необходимые данные. Произведена их предварительная подготовка и расчитаны столбцы со средним и суммарным значениями. Произведен анализ выбросов и аномалий.

###Сделанные наблюдения:
1. При наступлении пандемии Ковид19 произошло общее падение потребленя электроэнергии, что говорит о снижении промышленного производства.
2. Штаты с более сильной экономикой в период пандемии даже сумели нарастить промышленное производство, в то время как штаты со слабой экономикой в пандемию свои производства закрывали.
3. После начала пандемии Ковид19 потребление электроэнергии стало "ступенчатым", т.е. происходили резкие политические изменения: закрытие предприятий на карантин и возврат к работе предприятий.